### PR TITLE
Add cloudevents support to elixir contracts

### PR DESCRIPTION
**Pull request by CDimonaco**
- Add function to serialize an elixir contract to cloud event
- Add function to deserialize a raw cloud event into a contract struct

Both processes will involve contract validation.

We do that in order to encapsulate the cloud event details into the contracts library.

### CODE REVIEW

**Comment by CDimonaco** [comment]
Add validation for both direction when serializing/deserializing cloud event.

Schema embedded into priv directory and added to the source at compile time.

Changed makefile tasks in order to perform the schema copy like golang contracts.

File: elixir/lib/mix/tasks/generate.ex [hunk 1]
```diff
@@ -359,6 +379,53 @@ defmodule Mix.Tasks.Contracts.Generate do
...
 
+  @doc """
+    Generates the cloud events conversion functions for the contract
+  """
+  def trento_cloudevent_conversion_functions() do
+    quote do
+      @doc """
+        Serialize and return a module struct from a cloud event json
+      """
+      def serialize_from_cloud_event(json_cloud_event) do
```
Comment (arbulu89) [inline elixir/lib/mix/tasks/generate.ex:390]: Could we maybe add a signature to the method?
In this case it needs a binary, right?
Something like:
```
@spec serialize_from_cloud_event(json :: binary()) ::
          {:ok, t()} | {:error, any}
```

File: elixir/lib/mix/tasks/generate.ex [hunk 2]
```diff
@@ -359,6 +379,53 @@ defmodule Mix.Tasks.Contracts.Generate do
...
+        end
+      end
+
+      defp validate_with_schema(data) do
+        JsonSchema.validate(@event_type, data)
+      end
+
+      defp create_contract_from_cloud_event(%CloudEvent{data: data, type: @event_type}) do
+        with :ok <- validate_with_schema(data) do
+          __MODULE__.new(data)
```
Comment (arbulu89) [inline elixir/lib/mix/tasks/generate.ex:419]: Does this already do the `base64` conversion?
Just curiosity

File: elixir/lib/mix/tasks/generate.ex [hunk 3]
```diff
@@ -359,6 +379,53 @@ defmodule Mix.Tasks.Contracts.Generate do
...
+      def serialize_from_cloud_event(json_cloud_event) do
+        with {:ok, event} <- Cloudevents.from_json(json_cloud_event) do
+          create_contract_from_cloud_event(event)
+        end
+      end
+
+      @doc """
+        Serialize and return a cloud event json from a contract struct
+      """
+      def serialize_to_cloud_event(contract) do
```
Comment (arbulu89) [inline elixir/lib/mix/tasks/generate.ex:399]: What is the signature in this case? Does it receive a binary or some already predifined struct?

**Review by arbulu89 — APPROVED** [review]
Hey @CDimonaco ,
It looks good to me.
Could we maybe somehow change the code so the docstrings are correctly generated? I see some new line charaters etc in the generated code methods hehe

File: elixir/lib/mix/tasks/generate.ex (hunk 2 above)
Comment (CDimonaco) [reply elixir/lib/mix/tasks/generate.ex:419]: Nope, after our discussion I think that it's better if we skip the encoding, I would change that in the golang generation too after we merge this PR :D

File: elixir/lib/mix/tasks/generate.ex (hunk 3 above)
Comment (CDimonaco) [reply elixir/lib/mix/tasks/generate.ex:399]: It receives the __MODULE__ the struct of the contract we are generating, I will update them with the signatures

**Comment by CDimonaco** [comment]
Signature added to cloud events function in generated modules

**Pull request merged**